File: backend/app/api/routes/campaigns.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database.session import get_db
from app.schemas.campaign import CampaignCreate, CampaignResponse, CampaignUpdate, LeadResponse
from app.schemas.common import PaginatedResponse
from app.services.campaign import CampaignService

router = APIRouter()


@router.get("", response_model=PaginatedResponse[CampaignResponse])
@router.get("/", response_model=PaginatedResponse[CampaignResponse])
async def get_campaigns(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=100),
    status: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
):
    items, total = await CampaignService.get_campaigns(
        db, skip=skip, limit=limit, status=status
    )
    return PaginatedResponse(
        items=items,
        total=total,
        page=skip // limit + 1 if limit > 0 else 1,
        size=limit,
        pages=(total + limit - 1) // limit if limit > 0 else 1,
    )


@router.get("/{campaign_id}", response_model=CampaignResponse)
async def get_campaign(campaign_id: str, db: AsyncSession = Depends(get_db)):
    return await CampaignService.get_campaign(db, campaign_id)


@router.post("", response_model=CampaignResponse, status_code=status.HTTP_201_CREATED)
@router.post("/", response_model=CampaignResponse, status_code=status.HTTP_201_CREATED)
async def create_campaign(
    campaign_in: CampaignCreate, db: AsyncSession = Depends(get_db)
):
    return await CampaignService.create_campaign(db, campaign_in)


@router.patch("/{campaign_id}", response_model=CampaignResponse)
async def update_campaign(
    campaign_id: str, campaign_in: CampaignUpdate, db: AsyncSession = Depends(get_db)
):
    return await CampaignService.update_campaign(db, campaign_id, campaign_in)


@router.delete("/{campaign_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_campaign(campaign_id: str, db: AsyncSession = Depends(get_db)):
    await CampaignService.delete_campaign(db, campaign_id)


@router.get("/{campaign_id}/leads", response_model=PaginatedResponse[LeadResponse])
async def get_campaign_leads(
    campaign_id: str,
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    search: Optional[str] = None,
    qualification_status: Optional[str] = None,
    sort_by: Optional[str] = Query("created_at"),
    sort_order: Optional[str] = Query("desc"),
    db: AsyncSession = Depends(get_db),
):
    """
    Get paginated real leads belonging to this campaign from the database.
    """
    items, total = await CampaignService.get_campaign_leads(
        db=db,
        campaign_id=campaign_id,
        skip=skip,
        limit=limit,
        search=search,
        qualification_status=qualification_status,
        sort_by=sort_by,
        sort_order=sort_order,
    )
    return PaginatedResponse(
        items=items,
        total=total,
        page=skip // limit + 1 if limit > 0 else 1,
        size=limit,
        pages=(total + limit - 1) // limit if limit > 0 else 1,
    )


@router.get("/{campaign_id}/leads/export")
async def export_campaign_leads(
    campaign_id: str,
    format: str = Query("csv", regex="^(csv|json)$"),
    db: AsyncSession = Depends(get_db),
):
    """
    Export all stored leads for this campaign in CSV or JSON format.
    """
    content, media_type = await CampaignService.export_campaign_leads(
        db=db, campaign_id=campaign_id, export_format=format
    )
    filename = f"campaign_{campaign_id}_leads.{format}"
    return Response(
        content=content,
        media_type=media_type,
        headers={"Content-Disposition": f'attachment; filename="{filename}"'}
    )

File: frontend/src/lib/api/campaigns.ts
import { apiClient } from "./client";

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface BackendLead {
  id: string;
  campaign_id: string;
  platform: string;
  username: string;
  full_name?: string | null;
  business_name?: string | null;
  bio?: string | null;
  followers: number;
  following: number;
  website?: string | null;
  email?: string | null;
  phone?: string | null;
  city?: string | null;
  state?: string | null;
  country?: string | null;
  category?: string | null;
  profile_url?: string | null;
  profile_image?: string | null;
  source: string;
  qualification_status: string;
  created_at: string;
  updated_at: string;
}

export interface Campaign {
  id: string;
  name: string;
  platform: string;
  status: string;
  description?: string | null;
  config?: Record<string, any> | null;
  scraper_type?: string | null;
  scraperType?: string | null;
  lead_count?: number;
  created_at: string;
  updated_at: string;
}

export type CampaignCreate = Omit<Campaign, "id" | "created_at" | "updated_at" | "status" | "lead_count"> & { 
  status?: string; 
  config?: Record<string, any>;
};
export type CampaignUpdate = Partial<CampaignCreate>;

export interface LeadQueryParams {
  skip?: number;
  limit?: number;
  search?: string;
  qualification_status?: string;
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

export const campaignsApi = {
  getCampaigns: async (params?: { skip?: number; limit?: number; status?: string; platform?: string }) => {
    const { data } = await apiClient.get<PaginatedResponse<Campaign>>("/campaigns", { params });
    return data;
  },
  
  getCampaign: async (id: string) => {
    const { data } = await apiClient.get<Campaign>(`/campaigns/${id}`);
    return data;
  },
  
  createCampaign: async (campaign: CampaignCreate) => {
    const { data } = await apiClient.post<Campaign>("/campaigns", campaign);
    return data;
  },
  
  updateCampaign: async (id: string, updates: CampaignUpdate) => {
    const { data } = await apiClient.patch<Campaign>(`/campaigns/${id}`, updates);
    return data;
  },
  
  deleteCampaign: async (id: string) => {
    await apiClient.delete(`/campaigns/${id}`);
  },

  getCampaignLeads: async (campaignId: string, params?: LeadQueryParams) => {
    const { data } = await apiClient.get<PaginatedResponse<BackendLead>>(`/campaigns/${campaignId}/leads`, {
      params,
    });
    return data;
  },

  getExportUrl: (campaignId: string, format: "csv" | "json" = "csv") => {
    const baseURL = apiClient.defaults.baseURL || "http://localhost:8000/api";
    return `${baseURL}/campaigns/${campaignId}/leads/export?format=${format}`;
  },
};
